File: src/engines/Examples/Basic/01-GroupsAndObsSpaces.cpp
/*
 * (C) Copyright 2020-2021 UCAR
 *
 * This software is licensed under the terms of the Apache Licence Version 2.0
 * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
 */
/*! \defgroup ioda_cxx_ex Examples
 * \brief C++ Usage Examples
 * \ingroup ioda_cxx
 *
 * @{
 * \dir Basic
 * \brief Basic C++ ioda-engines usage tutorials
 *
 * \defgroup ioda_cxx_ex_1 Ex 1: Groups and ObsSpaces
 * \brief Group manipulation using the C++ interface
 * \see 01-GroupsAndObsSpaces.cpp for the example.
 *
 * @{
 *
 * \file 01-GroupsAndObsSpaces.cpp
 * \brief First example showing how to make a group (or an ObsSpace).
 *
 * IODA is the Interface for Observation Data Access.
 *
 *  The objective of the IODA is to provide uniform access to
 *  observation data across the whole forecasting chain from
 *  observation pre-processing to data assimilation to diagnostics.
 *
 *  IODA isolates the scientific code from the underlying data
 *  structures holding the data. Of course, any user needs to have
 *  a basic understanding of how the data are laid out. How are the
 *  data grouped, how can you access the data, how can you
 *  interpret variable names, and how can you read dimensions
 *  and meta data?
 *
 *  Data in IODA are stored in a structure Groups, Variables, and
 *  Attributes. A Group is like a folder. It is a logical
 *  collection of Variables and Attributes that describes some
 *  portion of the overall data. A Variable stores bulk data.
 *  An Attribute stores smaller quantities of metadata, and can
 *  be attached to either a Group or a Variable.
 *
 *  In ioda-engines, we separate out how the data are stored
 *  (the backend engine) from how an end user can access it (in
 *  the frontend). Different backends all provide the same general
 *  interface, but may support slightly different features and will
 *  have different performance characteristics.
 *
 *  This example shows you how to create Groups. It creates an
 *  HDF5 file, "Example-01.hdf5" using the HDF5 backend. Later examples
 *  will use groups to store and read data.
 *
 * \author Ryan Honeyager (honeyage@ucar.edu)
 **/

#include <iostream>  // We want I/O.
#include <string>    // We want strings
#include <vector>    // We want vectors

#include "ioda/Engines/Factory.h"  // Used to kickstart the Group engine.
#include "ioda/Group.h"            // We are manipulating ioda::Groups.

int main(int argc, char** argv) {
  using namespace ioda;  // All of the ioda functions are here.
  using std::cerr;
  using std::endl;
  using std::string;
  using std::vector;
  try {
    // We want to open a new file, backed by the default engine (HDF5).
    // We open this file as a root-level Group.
    Group grpFromFile = Engines::constructFromCmdLine(argc, argv, "Example-01.hdf5");
    // Note: After you build and run this example, you can view the contents of this
    // HDF5 file with either the "h5dump" or "ncdump" commands.
    // Since we are only using a subset of HDF5's full feature set, netCDF conveniently
    // recognizes this as a valid netCDF-4 file!

    // The only time that you need to be concerned about the
    // backend (HDF5) is when you create or open a root-level Group.
    // All Variables and Attributes within a Group transparently
    // use the same backend.

    // Groups can contain other Groups!
    // To create a new group, use the .create() method.
    // The new group is a child group of the object that is used to
    // create it. It shares the same underlying backend as its parent.
    Group g1 = grpFromFile.create("g1");
    Group g2 = grpFromFile.create("g2");

    // Groups can form a tree structure.
    Group g3 = g1.create("g3");
    Group g4 = g3.create("g4");
    Group g5 = g4.create("g5");
    Group g6 = g4.create("g6");
    Group g8 = g6.create("g7/g8");

    // Your tree looks like this:
    //
    // / - g1 - g3 - g4 - g5
    //   |                |
    //   g2               g6 - g7 - g8
    //

    // Besides creating Groups, we can also check if a particular
    // group exists, list them and open them.

    // Checking existance
    bool g3exists = g1.exists("g3");
    if (!g3exists) throw;  // jedi_throw.add("Reason", "g3 does not exist.");

    // Nesting
    // We can use '/' as a path separator.
    bool g4exists = g1.exists("g3/g4");
    if (!g4exists) throw;  // jedi_throw.add("Reason", "g4 does not exist.");

    // Listing the groups contained within a group.
    // The .list() function returns a vector of strings
    // listing all immediate (one-level) child groups.
    vector<string> g3children = g3.list();  // Should be { "g4" }.
    if (g3children.size() != 1)
      throw;                                /* jedi_throw
                 .add("Reason", "g3 contents are unexpected.")
                 .add("size", g3children.size()); */
    vector<string> g4children = g4.list();  // Should be { "g5", "g6" }.
    if (g4children.size() != 2)
      throw; /* jedi_throw
.add("Reason", "g4 contents are unexpected.")
.add("size", g4children.size()); */

    // When an error is encountered, you'll notice that
    // we did a "throw jedi_throw" statement. This throws an error
    // exception that is implemented in the jedi code.
    // "jedi_throw" is a macro that tags the current source file,
    // current function, and the line # in the current file, and
    // creates an "xThrow" object. You can then use the
    // .add(key, value) function to add in diagnostic information.
    // You can add any information that you could write using an
    // iostream like cout.
    // So, you can pass values like strings and numbers without
    // difficulty. Complex objects like vectors or sets would
    // need more work.

    // Opening groups
    // This is also really easy. Use the .open function.
    // It also obeys nesting criteria, and throws on an error.
    Group opened_g3 = g1.open("g3");
    Group opened_g6 = opened_g3.open("g4/g6");

    // Groups g3 and opened_g3 point to the same object.
    // Groups g6 and opened_g6 also point to the same object.
    // Any changes that you make in one of these groups will
    // be instantly visible to the other.

    // Note: we make no guarantees about concurrent access using
    // threads. That is a detail left up to the backend, and is
    // an area of future work.

    // What about closing groups?
    // These Group objects can go out of scope, and they release
    // their resource locks when they destruct. So, there is no
    // specific close method.
    // If you _really_ want to close an object, just reassign it.
    opened_g3 = Group{};  // opened_g3 now points elsewhere.

    // If all references to a specific backend instance are closed,
    // then it is released and does its cleanup tasks.

    // What about Attributes and Variables?
    // See tutorial 2 for Attributes.
    // Variables are covered in tutorials 3-7.

    // Thanks for reading!
  } catch (const std::exception& e) {
    cerr << "An error occurred.\n\n" << e.what() << endl;
    return 1;
  }
  return 0;
}

/// @}
/// @}

File: src/engines/Examples/Basic/01-GroupsAndObsSpaces.f90
!
! (C) Copyright 2022 UCAR
!
! This software is licensed under the terms of the Apache Licence Version 2.0
! which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.

!>  Group manipulation using the Fortran interface. This example parallels the C++ example.

program ioda_fortran_01_groupsandobsspaces
    use, intrinsic :: iso_c_binding
    use, intrinsic :: iso_fortran_env
    use :: ioda_vecstring_mod
    use :: ioda_group_mod
    use :: ioda_engines_mod
    implicit none

    type(ioda_group) :: groot, g1, g2, g3, g4, g5, g6, g7, g8, reopened_g3
    type(ioda_vecstring) :: listed_groups_g3,listed_groups_g4
    integer(int64) :: nsz
    integer(int64) :: msz
    character(len=*),parameter :: root_file = 'Example-01.hdf5'
    ! Create a file
    !grpFromFile = engines%obsstore%createRootGroup()
    integer :: create_mode
    
    create_mode = 1             

    call ioda_engines_construct_from_command_line(groot,root_file) 
    
    nsz = 2
    msz = 5
    call groot%create(nsz,'g1',g1)
    call groot%create(nsz,'g2',g2)
    call g1%create(nsz,'g3',g3)
    call g3%create(nsz,'g4',g4)
    call g4%create(nsz,'g5',g5)
    call g4%create(nsz,'g6',g6)
    call groot%create(msz,'g7/g8',g7)
 
    ! Check for existence of groups
    if (g1%exists(nsz,'g3') .eq. 0) then
        write(error_unit,*) '/g1/g3 does not exist'
        stop -1
    end if

    if (groot%exists(msz,'g7/g8') .eq. 0) then
        write(error_unit,*) '/groo/g7/g8 does not exist'
        stop -1
    end if

    call ioda_vecstring_init(listed_groups_g3)	
    ! Listing groups contained within a group. See VecString example for full usage.
    call g3%list(listed_groups_g3)
    if (listed_groups_g3%size() /= 1) then
        write(error_unit,*) 'Unexpected number of child groups for g3.'
        stop -1
    end if
    
    call ioda_vecstring_init(listed_groups_g4)	
    call g4%list(listed_groups_g4)
    if (listed_groups_g4%size() /= 2) then
        write(error_unit,*) 'Unexpected number of child groups for g3.'
        stop -1    
    end if         

    ! Open groups
    call g3%open(nsz,'g4',reopened_g3)

end program	
